File: natality/prep.jl
using DataFrames, CSV, Printf, CodecZlib

# Path to the data files
pa = "/home/kshedden/mynfs/data/Teaching/natality"

# Create a long form version of the births
dl = []
for y = 2016:2020
    fn = joinpath(pa, @sprintf("%4d.txt.gz", y))
    tm = Dict("County Code" => String, "Births" => Float64)
    da = open(fn) do io
        CSV.read(io, DataFrame, delim = "\t", types = tm, silencewarnings = true)
    end
    da = da[:, ["County", "County Code", "Births"]]
    da = da[completecases(da), :]
    da[:, :year] .= y
    push!(dl, da)
end
births = vcat(dl...)
births = rename(births, "County Code" => "FIPS")

# Subset the demographics file to 2016
if false
    f = joinpath(pa, "us.1990_2020.19ages.txt.gz")
    g = joinpath(pa, "2016ages.txt.gz")
    open(GzipCompressorStream, g, "w") do out
        open(GzipDecompressorStream, f) do inp
            for line in eachline(inp)
                if startswith(line, "2016")
                    write(out, line)
                    write(out, "\n")
                end
            end
        end
    end
end

# Read the demographics for 2016.  It is a fixed-width format
# file.  Unfortunately Julia does not currently have a good
# fixed width file reader so we have to do a lot of tedious
# processing here.
x = [1, 5, 7, 9, 12, 14, 15, 16, 17, 19, 27]
cs = [(x[i], x[i+1] - 1) for i = 1:length(x)-1]
demog = (
    Year = Int[],
    State = String[],
    StateFIPS = String[],
    CountyFIPS = String[],
    Registry = String[],
    Race = String[],
    Origin = String[],
    Sex = String[],
    Age = String[],
    Population = Float64[],
)
open(GzipDecompressorStream, joinpath(pa, "2016ages.txt.gz")) do io
    for line in eachline(io)
        for j = 1:length(demog)
            v = line[cs[j][1]:cs[j][2]]
            T = eltype(demog[j])
            if T <: AbstractString
                push!(demog[j], v)
            elseif T <: Number
                push!(demog[j], parse(T, v))
            end
        end
    end
end
demog = DataFrame(demog)

# Create a FIPS code that matches the FIPS code in the birth data
demog[:, :FIPS] = ["$x$y" for (x, y) in zip(demog.StateFIPS, demog.CountyFIPS)]
demog = demog[:, [:FIPS, :Race, :Origin, :Sex, :Age, :Population]]

# Recode some variables to more interpretable text labels
demog[:, :Sex] = replace(demog[:, :Sex], "1" => "M", "2" => "F")
demog[:, :Origin] = replace(demog[:, :Origin], "0" => "N", "1" => "H")
demog[:, :Race] = replace(demog[:, :Race], "1" => "W", "2" => "B", "3" => "N", "4" => "A")

# The overall population per county
pop = combine(groupby(demog, :FIPS), :Population => sum)
pop = rename(pop, :Population_sum => :Population)

# Pivot to put the age bands in the columns
demog[:, :group] = [
    "$(w)_$(x)_$(y)_$(z)" for (w, x, y, z) in
    zip(demog[:, :Race], demog[:, :Origin], demog[:, :Sex], demog[:, :Age])
]
demog = select(demog, Not([:Race, :Origin, :Sex, :Age]))
demog = unstack(demog, :group, :Population)

na = names(demog)[2:end]
na = [split(x, "_") for x in na]
na = DataFrame(
    Race = [x[1] for x in na],
    Origin = [x[2] for x in na],
    Sex = [x[3] for x in na],
    Age = [x[4] for x in na],
)

# Get the Rural/Urban Continuity Codes (RUCC)
rucc = open("rucc2013.csv.gz") do io
    CSV.read(io, DataFrame, types = Dict(:FIPS => String))
end

File: natality/prep.py
# Prepare the birth count and other related data for analysis.

import pandas as pd
import numpy as np
import os, gzip
from pathlib import Path

# Path to the data files
pa = Path("/home/kshedden/data/Teaching/natality")

# Create a long form version of the birth data.
dl = []
for y in range(2011, 2021):
    bname = "%4d.txt.gz" % y
    da = pd.read_csv(pa / bname, delimiter="\t",
                     dtype={"County Code": object})
    da = da[["County", "County Code", "Births"]]
    da["year"] = y
    dl.append(da)
births = pd.concat(dl)
births = births.rename({"County Code": "FIPS"}, axis=1)
births["Births"] = pd.to_numeric(births.Births, errors="coerce")
births = births.dropna()
births = births[~births["County"].str.contains("Unidentified")]

# Subset the demographics file to 2016.  This is a large file
# and takes time, so only run this once and save the result
# for future runs.
if False:
    f = pa / "us.1990_2023.20ages.adjusted.txt.gz"
    g = pa / "2016ages.txt.gz"
    with gzip.open(g, "w") as out:
        with gzip.open(f) as inp:
            for line in inp:
                if line.startswith(b"2016"):
                    out.write(line)

# Read the demographics for 2016.  It is a fixed-width format
# file.
x = [1, 5, 7, 9, 12, 14, 15, 16, 17, 19, 27]
cs = [(x[i]-1, x[i+1]-1) for i in range(len(x)-1)]
with gzip.open(pa / "2016ages.txt.gz") as io:
    demog = pd.read_fwf(io, colspecs=cs, header=None)
demog.columns = ["Year", "State", "StateFIPS", "CountyFIPS", "Registry",
                 "Race", "Origin", "Sex", "Age", "Population"]

# Create a FIPS code that matches the FIPS code in the birth data
demog["FIPS"] = ["%02d%03d" % (x, y) for (x, y) in zip(demog.StateFIPS, demog.CountyFIPS)]
demog = demog[["FIPS", "Race", "Origin", "Sex", "Age", "Population"]]

# Recode some variables to more interpretable text labels
# See http://seer.cancer.gov/popdata/popdic.html for code information
demog["Sex"] = demog["Sex"].replace([2, 1], ["F", "M"]) # Female/Male
demog["Origin"] = demog["Origin"].replace([0, 1], ["N", "H"]) # Non-Hispanic/Hispanic
demog["Race"] = demog["Race"].replace([1, 2, 3, 4], ["W", "B", "N", "A"]) # White/Black/Native/Asian

# The overall population per county
pop = demog.groupby("FIPS")["Population"].sum().reset_index()

# Pivot to put the age bands in the columns
demog = demog.pivot(index="FIPS", columns=["Race", "Origin", "Sex", "Age"], values="Population")

# Age group labels
age_groups = ["0", "1-4", "5-9", "10-14", "15-19", "20-24", "25-29", "30-34", "35-39",
              "40-44", "45-49", "50-54", "55-59", "60-64", "65-69", "70-74", "75-79",
              "80-84", "85-89", "90+"]

na = demog.columns.tolist()
demog.columns = ["%s_%s_%s_%d" % tuple(x) for x in na]

# Replace missing demographic values with 0.
demog = demog.fillna(0)

# Get the Rural/Urban Continuity Codes (RUCC)
rucc = pd.read_excel(pa / "ruralurbancodes2013.xls", sheet_name=None)
rucc = rucc["Rural-urban Continuum Code 2013"]
rucc = rucc[["FIPS", "RUCC_2013"]]
rucc["FIPS"] = ["%05d" % x for x in rucc.FIPS]

# Get the ADI values, need to aggregate from zip code to county code
fname = "US_2022_ADI_Census_Block_Group_v4_0_1.csv.gz"
dt = {"FIPS": str, "ADI_NATRANK": str}
adi = pd.read_csv(pa / fname, usecols=["FIPS", "ADI_NATRANK"], dtype=dt)
adi["ADI_NATRANK"] = pd.to_numeric(adi["ADI_NATRANK"], errors="coerce")
adi["FIPS5"] = [x[0:5] for x in adi["FIPS"]]
adi = adi.groupby("FIPS5")["ADI_NATRANK"].agg("median")
